tests/testthat/test-teal_data_module-eval_code.R

```diff
@@ -0,0 +1,169 @@
+        within(iris, id <- seq_len(NROW(Species)))
+      )
+    }
+  ) %>%
+    testthat::expect_no_error()
+})
+
+testthat::test_that("eval_code.teal_data_module will execute several times until error", {
+  tdm <- teal_data_module(
+    ui = function(id) div(),
+    server = function(id) {
+      shiny::moduleServer(id, function(input, output, session) {
+        shiny::reactive(teal.data::teal_data(IRIS = iris))
+      })
+    }
+  )
+
+  tdm2 <- eval_code(tdm, quote(stop_me <- FALSE)) %>%
+    eval_code("stopifnot(previous_error = stop_me)")
+
+  shiny::testServer(
+    app = tdm2$server,
+    expr = {
+      testthat::expect_s3_class(td(), "qenv.error")
+      testthat::expect_match(td()$message, "previous_error.*when evaluating qenv code")
+    }
+  ) %>%
+    testthat::expect_no_error()
+})
+
+testthat::test_that("eval_code.teal_data_module throws error when original teal_data_module result is not reactive", {
+  tdm <- teal_data_module(
+    ui = function(id) div(),
+    server = function(id) {
+      shiny::moduleServer(id, function(input, output, session) {
+        "I am not reactive, I am a string"
+      })
+    }
+  )
+
+  tdm2 <- eval_code(tdm, "1 + 1")
+
+  shiny::testServer(
+    app = tdm2$server,
+    expr = {}
+  ) %>%
+    testthat::expect_error("The `teal_data_module` must return a reactive expression.")
+})
+
+testthat::test_that("eval_code.teal_data_module propagates qenv error from the original/first call", {
+  tdm <- teal_data_module(
+    ui = function(id) div(),
+    server = function(id) {
+      shiny::moduleServer(id, function(input, output, session) {
+        reactive(
+          teal_data(IRIS = iris) %>% within("non_existing_var + 1")
+        )
+      })
+    }
+  )
+
+  tdm2 <- eval_code(tdm, "IRIS$const <- 1 + 1")
+
+  shiny::testServer(
+    app = tdm2$server,
+    expr = {
+      testthat::expect_s3_class(
+        td(),
+        "qenv.error"
+      )
+    }
+  ) %>%
+    testthat::expect_no_error()
+})
+
+testthat::test_that("eval_code.teal_data_module handles an arbitrary object (other than `teal_data` or `qenv.error`)", {
+  tdm <- teal_data_module(
+    ui = function(id) div(),
+    server = function(id) {
+      shiny::moduleServer(id, function(input, output, session) {
+        reactive(list())
+      })
+    }
+  )
+
+  tdm2 <- eval_code(tdm, "1 + 1")
+
+  shiny::testServer(
+    app = tdm2$server,
+    expr = {
+      testthat::expect_identical(
+        td(),
+        list()
+      )
+    }
+  ) %>%
+    testthat::expect_no_error()
+})
+
+testthat::test_that("eval_code.teal_data_module handles a `NULL` result", {
+  tdm <- teal_data_module(
+    ui = function(id) div(),
+    server = function(id) {
+      shiny::moduleServer(id, function(input, output, session) {
+        reactive(NULL)
+      })
+    }
+  )
+
+  tdm2 <- eval_code(tdm, "1 + 1")
+
+  shiny::testServer(
+    app = tdm2$server,
+    expr = {
+      testthat::expect_null(td())
+    }
+  ) %>%
+    testthat::expect_no_error()
```

Comment:
Please don't ever use `%>%` in the developer code. Should be only in the vignettes/docs

Reply:
Ok, I used them in testthat as they are well integrated in testthat https://testthat.r-lib.org/news/index.html?q=pipe#new-expectations-1-0-0 and help create clean tests.

I'll remove them though if that is the standard for `{teal}`

Reply:
Done in b1625ba52 (also removed a native pipe in tests that was useless)